src/java.base/share/classes/java/lang/LazyConstant.java

```diff
@@ -0,0 +1,298 @@
+ * object will never change once it is initialized. However, the referenced object itself
+ * may or may not be mutable. Hence, immutability can only be guaranteed at the first
+ * initial, shallow level.
+ * <p>
+ * Consider the following example where a lazy constant field "{@code logger}" is a
+ * shallowly immutable holder of an object of type {@code Logger}:
+ *
+ * {@snippet lang = java:
+ * public class Component {
+ *
+ *    // Creates a new uninitialized lazy constant
+ *    private final LazyConstant<Logger> logger =
+ *            // @link substring="of" target="#of" :
+ *            LazyConstant.of( () -> Logger.create(Component.class) );
+ *
+ *    public void process() {
+ *        logger.get().info("Process started");
+ *        // ...
+ *    }
+ * }
+ *}
+ * <p>
+ * Initially, the lazy constant (and its contents) is <em>not initialized</em>.
+ * When {@code logger.get()} is first invoked, it evaluates the computing function and
+ * initializes the constant to the result; the result is then returned to the client.
+ * Hence, {@linkplain #get()} guarantees that the constant is <em>initialized</em> before
+ * it returns, barring any exceptions.
+ * <p>
+ * Furthermore, {@linkplain #get()} guarantees that, out of several threads trying to
+ * invoke the computing function simultaneously, only one is ever selected for
+ * computation. This property is crucial as evaluation of the computing function may have
+ * side effects, for example, the call above to {@code Logger.create()} may result in
+ * storage resources being prepared.
+ *
+ * <h2 id="exception-handling">Exception handling</h2>
+ * If the computing function returns {@code null}, a {@linkplain NullPointerException}
+ * is thrown. Hence, a lazy constant can never hold a {@code null} value. Clients who
+ * want to use a nullable constant can wrap the value into an {@linkplain Optional} holder.
+ * <p>
+ * If the computing function recursively invokes itself (directly or indirectly via
+ * the lazy constant), an {@linkplain IllegalStateException} is thrown, and the lazy
+ * constant is not initialized.
+ * <p>
+ * If the computing function throws any unchecked {@linkplain Exception} or
+ * {@linkplain Error}, that {@linkplain Throwable} is propagated to the caller, and the
+ * lazy constant remains uninitialized. In other words, upon an unsuccessful invocation of
+ * the computing function, neither a constant, the exception, nor the fact that
+ * an exception was thrown is ever stored in the lazy constant.
+ *
+ * <h2 id="composition">Composing lazy constants</h2>
+ * A lazy constant can depend on other lazy constants, forming a dependency graph
+ * that can be lazily computed but where access to individual elements can still be
+ * performant. In the following example, a single {@code Foo} and a {@code Bar}
+ * instance (that is dependent on the {@code Foo} instance) are lazily created, both of
+ * which are held by lazy constants:
+ *
+ * {@snippet lang = java:
```

Comment:
If you wanted, you could reuse the `Component`  class -- that one already has a lazy constant field, so if you put the component itself in another lazy constant field you have your composition example.